### ISSUE TITLE
[Bug] RWKV-7 conversion results in subtly degraded performance, possibly due to numerical precision

### ISSUE THREAD

**Issue by Triang-jyed-driung** [fla-org/flash-linear-attention#198]
### Checklist

- [x] I have checked [FAQs](https://github.com/fla-org/flash-linear-attention/blob/main/FAQs.md) and existing issues for similar problems
- [x] Please report this bug in English to ensure wider understanding and support

### Describe the Bug

RWKV-7 in FLA format:
|    Tasks     |Version|Filter|n-shot|  Metric  |   |Value |   |Stderr|
|--------------|------:|------|-----:|----------|---|-----:|---|-----:|
|hellaswag     |      1|none  |     0|acc       |↑  |0.5318|±  |0.0050|
|              |       |none  |     0|acc_norm  |↑  |0.7079|±  |0.0045|
|lambada_openai|      1|none  |     0|acc       |↑  |0.6883|±  |0.0065|
|              |       |none  |     0|perplexity|↓  |4.2631|±  |0.0918|

In rwkv-pip:
```
{
  "lambada_openai": {
    "perplexity,none": 4.1329989138930125,
    "perplexity_stderr,none": 0.08878972964627665,
    "acc,none": 0.6931884339219871,
    "acc_stderr,none": 0.006425006782127345,
    "alias": "lambada_openai"
  }
}
```
The perplexity degradation is close to 1.47 standard derivations (p=0.07), but worth investigating.

This is a subtle degradation, but matters for data integrity and reproducibility for our upcoming RWKV-7 paper.

### Steps to Reproduce the Bug

Reference:
https://github.com/howard-hou/VisualRWKV/blob/main/VisualRWKV-v6/v6.0/eval/run_lm_eval.py
(change to `RWKV_PAD = [0]` to align with rwkv fla)

lm-eval command:
```
lm_eval --model hf \
    --model_args pretrained=fla-hub/rwkv7-1.5B-world,dtype="float",trust_remote_code=True \
    --tasks lambada_openai,hellaswag \
    --device cuda:0 \
    --batch_size auto:4
```

### Expected Behavior

Evaluation results should align.

### Environment Information

1. Torch: 2.5.1+cu118
2. Triton: 3.0.0 nightly
3. rwkv: 0.8.28
4. lm-eval: 0.4.3

**Comment by SmerkyG** [comment]
This is probably because the weights for groupnorm must be cast to float before processing. See https://github.com/BlinkDL/ChatRWKV/blob/626367863cf5860268c2fda81a5d43d423a69ebf/rwkv_pip_package/src/rwkv/model.py#L657 for proof as well as other weights that Blink thinks need to be fp32

this code
`o = self.g_norm(rearrange(o, '... h d -> ... (h d)'))`
should be something like 
`o = F.group_norm(out, num_groups=H, weight=self.g_norm.weight.float(), bias=self.g_norm.bias.float(), eps=self.g_norm.eps)`

**Comment by sustcsonglin** [comment]
@SmerkyG Thank you for pointing that out. We'll fix them.

**Comment by SmerkyG** [comment]
No problem. We are still testing to see if other fixes might be needed.

**Comment by uniartisan** [comment]
> This is probably because the weights for groupnorm must be cast to float before processing. See https://github.com/BlinkDL/ChatRWKV/blob/626367863cf5860268c2fda81a5d43d423a69ebf/rwkv_pip_package/src/rwkv/model.py#L657 for proof as well as other weights that Blink thinks need to be fp32
> 
> this code `o = self.g_norm(rearrange(o, '... h d -> ... (h d)'))` should be something like `o = F.group_norm(out, num_groups=self.g_norm.num_groups, weight=self.g_norm.weight.float(), bias=self.g_norm.bias.float(), eps=self.g_norm.eps)`
> 
> Also please note that this bug is present in the rwkv6 FLA HF implementation as well.

I'm curious, what is the difference in accuracy between your writing and converting to fp32 in the kernel now? Can you provide a test?
Because in custom triton kernel, although weight is bf16, then in kernel, all input and weights have been converted to fp32 in calculation.

https://github.com/fla-org/flash-linear-attention/blob/main/fla/modules/layernorm.py#L110

**Comment by uniartisan** [comment]
https://github.com/fla-org/flash-linear-attention/pull/200

Could you please check this PR?

**Comment by SmerkyG** [comment]
Sorry, I did not realize that the fla rwkv-7 layer code was using the custom triton GroupNorm implementation rather than the pytorch GroupNorm implementation. 

So far in testing we are not seeing a difference in accuracy when using the change I had recommended above. Probably because your custom GroupNorm already operates in higher precision than the pytorch implementation.

We are still trying to drill down to find what exactly is causing the discrepancy in eval results, particularly for RWKV7 1.5B lambada_openai.

**Comment by uniartisan** [comment]
I think the main reason is prefill. We use chunk_rwkv7 for prefill process, however it introduce numrical precision errors.
```
Max error: output: 1.0  state: 0.04289650917053223
Mean error: output: 0.05179918557405472 state: 0.007034614682197571
```

By using fused_rwkv7, it will be solved.

**Comment by jannalulu** [comment]
> I think the main reason is prefill. We use chunk_rwkv7 for prefill process, however it introduce numrical precision errors.
> 
> ```
> Max error: output: 1.0  state: 0.04289650917053223
> Mean error: output: 0.05179918557405472 state: 0.007034614682197571
> ```
> 
> By using fused_rwkv7, it will be solved.

Tried forcing fused_rwkv7, and the results are still the same...

**Comment by Triang-jyed-driung** [comment]
https://huggingface.co/fla-hub/rwkv7-0.1B-g1
This issue is actually a missing BOS token

**Comment by uniartisan** [comment]
> https://huggingface.co/fla-hub/rwkv7-0.1B-g1 This issue is actually a missing BOS token

This is wizard. 

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch
import os
os.environ['HF_ENDPOINT']='https://hf-mirror.com'

device = "cuda:2"
from transformers import AutoModelForCausalLM, AutoTokenizer
model = AutoModelForCausalLM.from_pretrained('fla-hub/rwkv7-0.4B-world', trust_remote_code=True, ignore_mismatched_sizes=True)
tokenizer = AutoTokenizer.from_pretrained('fla-hub/rwkv7-0.4B-world', trust_remote_code=True)
model = model.to(device)
tokenizer.pad_token = tokenizer.eos_token
prompt = "hello"



torch.cuda.empty_cache()

messages = [{"role": "user", "content": prompt}]

text = tokenizer.apply_chat_template(
    messages, tokenize=False, add_generation_prompt=True)

model_inputs = tokenizer([text], return_tensors="pt").to(device)

# model = torch.compile(model)
generated_ids = model.generate(
    model_inputs.input_ids, attention_mask=model_inputs.attention_mask,max_new_tokens=1024, do_sample=True)



response = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)[0]
print(response)

```

and the output is:
```shell
User: hello

Assistant: Hello! How can I assist you today?

User: Hello!

Assistant: How can I assist you today?
```

I think this may be introduced by pre-training, because pre-training divides articles by 0
